core/store/src/opener.rs

```diff
@@ -11,7 +93,22 @@ const STORE_PATH: &str = "data";
 ///     .home(neard_home_dir)
 ///     .open();
 /// ```
-pub struct StoreOpener<'a> {
+pub struct StoreOpener<'a, M = ()> {
```

Comment:
I have mildly strong preference for `dyn` over templates here. This is high-level dispatch code, so we don't care about performance here at all. At the same time, we do care about separate compilation, because slow compile times suck. I suggest going with the following API: 

```rust
pub struct StoreOpener<'a> {
    pub(crate) db: DBOpener<'a>,
    mode: Mode,
    migrator: Option<Box<dyn FnMut(&NodeStorage, DbVersion) -> anyhow::Result<()> + 'a>>,
}

impl<'a> StoreOpener<'a> {
    pub(crate) fn new(home_dir: &std::path::Path, config: &'a StoreConfig) -> Self {
        Self { db: DBOpener::new(home_dir, config), mode: Mode::ReadWrite, migrator: None }
    }

    pub fn with_migrator(
        self,
        migrator: impl FnMut(&NodeStorage, DbVersion) -> anyhow::Result<()> + 'a,
    ) -> Self {
        Self { migrator: Box::new(migrator), ..self }
    }
}
```

Reply:
The reason for generic code and why I didn’t go with `with_migrator` is that I wanted to generalise the code on error type.  That is, StoreOpener does not need to know what kind of error the migrator returns.  It just packages it in StoreOpenerError::MigrationError and passes it along.  With dyn I’d have to make the error part of the interface.  Maybe I am overthinking this?

Reply:
Yeah, I think, in practice, anyhow, io::Error, or even just String would work fine here. 

If the caller absolutely insists on getting a well-typed error, they can do something like this: 


```
let mut err: Option<MyMigrationError> = None;
StoreOpener::whith_migration(|| {
    migrate().map_err(|e| err = Some(e))
}).open();
If let Some(err) = err { }
```

That's how `std::fmt::Error` works I think.